### PR TITLE
[FEATURE] Cleanup and stable alpine

### PR DESCRIPTION
**Pull request by im-kulikov**
- alpine 3.7
- dep prune
- fixed Dockerfile

### CODE REVIEW

**Comment by im-kulikov** [comment]
@vishr @alexandreStein @SkyrisBactera @linki ping?

File: Dockerfile [hunk 1]
```diff
@@ -1,12 +1,21 @@
-FROM alpine:edge
+# Build image
+FROM golang:alpine AS builder
 MAINTAINER Vishal Rana <vr@labstack.com>
 
-ENV VERSION 0.3.7
+COPY . /go/src/github.com/labstack/armor
```
Comment (alexandreStein) [inline Dockerfile]: It doesn't work for me.
All external parts are missing:
```
Sending build context to Docker daemon  8.093MB
Step 1/10 : FROM golang:alpine AS builder
 ---> f421e93ece9c
Step 2/10 : MAINTAINER Vishal Rana <vr@labstack.com>
 ---> Using cache
 ---> b66c4ae6685a
Step 3/10 : COPY . /go/src/github.com/labstack/armor
 ---> 33cc5a3b5854
Step 4/10 : WORKDIR /go/src/github.com/labstack/armor
Removing intermediate container dc42cbcecb64
 ---> 87b84162cabd
Step 5/10 : RUN set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go
 ---> Running in a91cf64ab36b
+ export CGO_ENABLED=0
+ go build -v -o /go/bin/armor cmd/armor/main.go
plugin/plugin.go:10:2: cannot find package "github.com/Knetic/govaluate" in any of:
	/usr/local/go/src/github.com/Knetic/govaluate (from $GOROOT)
	/go/src/github.com/Knetic/govaluate (from $GOPATH)
cmd/armor/main.go:10:2: cannot find package "github.com/go-yaml/yaml" in any of:
	/usr/local/go/src/github.com/go-yaml/yaml (from $GOROOT)
	/go/src/github.com/go-yaml/yaml (from $GOPATH)
plugin/message.pb.go:16:8: cannot find package "github.com/golang/protobuf/proto" in any of:
	/usr/local/go/src/github.com/golang/protobuf/proto (from $GOROOT)
	/go/src/github.com/golang/protobuf/proto (from $GOPATH)
armor.go:7:2: cannot find package "github.com/labstack/echo" in any of:
	/usr/local/go/src/github.com/labstack/echo (from $GOROOT)
	/go/src/github.com/labstack/echo (from $GOPATH)
plugin/body_limit.go:5:2: cannot find package "github.com/labstack/echo/middleware" in any of:
	/usr/local/go/src/github.com/labstack/echo/middleware (from $GOROOT)
	/go/src/github.com/labstack/echo/middleware (from $GOPATH)
armor.go:8:2: cannot find package "github.com/labstack/gommon/color" in any of:
	/usr/local/go/src/github.com/labstack/gommon/color (from $GOROOT)
	/go/src/github.com/labstack/gommon/color (from $GOPATH)
armor.go:9:2: cannot find package "github.com/labstack/gommon/log" in any of:
	/usr/local/go/src/github.com/labstack/gommon/log (from $GOROOT)
	/go/src/github.com/labstack/gommon/log (from $GOPATH)
http/http.go:14:2: cannot find package "github.com/mitchellh/go-homedir" in any of:
	/usr/local/go/src/github.com/mitchellh/go-homedir (from $GOROOT)
	/go/src/github.com/mitchellh/go-homedir (from $GOPATH)
plugin/plugin.go:14:2: cannot find package "github.com/mitchellh/mapstructure" in any of:
	/usr/local/go/src/github.com/mitchellh/mapstructure (from $GOROOT)
	/go/src/github.com/mitchellh/mapstructure (from $GOPATH)
plugin/plugin.go:15:2: cannot find package "github.com/valyala/fasttemplate" in any of:
	/usr/local/go/src/github.com/valyala/fasttemplate (from $GOROOT)
	/go/src/github.com/valyala/fasttemplate (from $GOPATH)
http/http.go:15:2: cannot find package "golang.org/x/crypto/acme/autocert" in any of:
	/usr/local/go/src/golang.org/x/crypto/acme/autocert (from $GOROOT)
	/go/src/golang.org/x/crypto/acme/autocert (from $GOPATH)
The command '/bin/sh -c set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go' returned a non-zero code: 1
```

I would replace it by: `RUN go get -u github.com/labstack/armor`.
But this is just a guess.

Comment (im-kulikov) [reply Dockerfile]: hm.. I'll try, maybe some changes was made in vendor

**Comment by im-kulikov** [comment]
```
→ docker build .
Sending build context to Docker daemon  14.74MB
Step 1/10 : FROM golang:alpine AS builder
alpine: Pulling from library/golang
1160f4abea84: Already exists 
4d49542c61a4: Pull complete 
3ee103c86f60: Pull complete 
9a56d2eb1eed: Pull complete 
67e0ebed9a3b: Pull complete 
194910951a14: Pull complete 
Digest: sha256:354be5853ea170e6f8bf3e258154e10ba0ed03f909d8be8625faf61592c515c8
Status: Downloaded newer image for golang:alpine
 ---> f421e93ece9c
Step 2/10 : MAINTAINER Vishal Rana <vr@labstack.com>
 ---> Running in 972088f9e07f
Removing intermediate container 972088f9e07f
 ---> 055ca53792fa
Step 3/10 : COPY . /go/src/github.com/labstack/armor
 ---> 907b84593d4a
Step 4/10 : WORKDIR /go/src/github.com/labstack/armor
Removing intermediate container 7bdb9411085a
 ---> a6203df54ad2
Step 5/10 : RUN set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go
 ---> Running in 1662209663b7
+ export CGO_ENABLED=0
+ go build -v -o /go/bin/armor cmd/armor/main.go
github.com/labstack/armor/vendor/github.com/go-yaml/yaml
net
github.com/labstack/armor/vendor/github.com/mattn/go-isatty
github.com/labstack/armor/vendor/github.com/valyala/bytebufferpool
github.com/labstack/armor/vendor/github.com/Knetic/govaluate
github.com/labstack/armor/vendor/github.com/golang/protobuf/proto
github.com/labstack/armor/vendor/github.com/labstack/gommon/bytes
github.com/labstack/armor/vendor/github.com/labstack/gommon/random
github.com/labstack/armor/vendor/github.com/mattn/go-colorable
github.com/labstack/armor/vendor/github.com/mitchellh/mapstructure
github.com/labstack/armor/vendor/github.com/valyala/fasttemplate
github.com/labstack/armor/vendor/github.com/labstack/gommon/color
github.com/labstack/armor/vendor/github.com/labstack/armor/util
github.com/labstack/armor/vendor/github.com/mitchellh/go-homedir
github.com/labstack/armor/vendor/github.com/labstack/gommon/log
crypto/x509
vendor/golang_org/x/net/lex/httplex
vendor/golang_org/x/net/proxy
net/textproto
mime/multipart
crypto/tls
github.com/labstack/armor/vendor/github.com/dgrijalva/jwt-go
net/http/httptrace
net/http
github.com/labstack/armor/vendor/golang.org/x/crypto/acme
net/http/httputil
github.com/labstack/armor/vendor/golang.org/x/crypto/acme/autocert
github.com/labstack/armor/vendor/github.com/labstack/echo
github.com/labstack/armor/vendor/github.com/labstack/armor
github.com/labstack/armor/vendor/github.com/labstack/echo/middleware
github.com/labstack/armor/vendor/github.com/labstack/armor/plugin
github.com/labstack/armor/vendor/github.com/labstack/armor/http
command-line-arguments
Removing intermediate container 1662209663b7
 ---> 122ef7bf44c9
Step 6/10 : FROM scratch
 ---> 
Step 7/10 : WORKDIR /
 ---> Using cache
 ---> be37fe24befb
Step 8/10 : COPY --from=builder /go/bin/armor /armor
 ---> 4e5707d3074c
Step 9/10 : COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
 ---> de7d4bdc3f78
Step 10/10 : ENTRYPOINT ["/armor"]
 ---> Running in 1e7446d5265b
Removing intermediate container 1e7446d5265b
 ---> 89ca707e4e7d
Successfully built 89ca707e4e7d
```

@alexandreStein is there any problems?

File: Dockerfile (hunk 1 above)
Comment (im-kulikov) [reply Dockerfile]: > I can't run anything (mkdir) on your image.
> I maybe missing something...

If you need to do something like "mkdir" or other linux-tools, do that in your Dockerfile...

Comment (alexandreStein) [reply Dockerfile]: I'm doing it in the Dockerfile of the next image and it doesn't work:

```
docker build -t armor-test .
Sending build context to Docker daemon  13.82kB
Step 1/7 : FROM armor-pull
 ---> 43c99df892a0
Step 2/7 : RUN mkdir -p /etc/armor/certs &&   mkdir -p /var/test
 ---> Running in e690865f20f2
OCI runtime create failed: container_linux.go:296: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown
```

I maybe doing it wrong.

**Comment by alexandreStein** [comment]
It's working.

You asked for my logs:
```
docker build -t armor-pull .
Sending build context to Docker daemon  14.49MB
Step 1/10 : FROM golang:alpine AS builder
 ---> f421e93ece9c
Step 2/10 : MAINTAINER Vishal Rana <vr@labstack.com>
 ---> Running in 7f6980a97150
Removing intermediate container 7f6980a97150
 ---> 8e73fa37e6b6
Step 3/10 : COPY . /go/src/github.com/labstack/armor
 ---> 667d433ce467
Step 4/10 : WORKDIR /go/src/github.com/labstack/armor
Removing intermediate container 5c7b64c9576c
 ---> 5490d6ddc206
Step 5/10 : RUN set -x     && export CGO_ENABLED=0     && go build -v -o /go/bin/armor cmd/armor/main.go
 ---> Running in 0e3c19dc9727
+ export CGO_ENABLED=0
+ go build -v -o /go/bin/armor cmd/armor/main.go
github.com/labstack/armor/vendor/github.com/go-yaml/yaml
net
github.com/labstack/armor/vendor/github.com/mattn/go-isatty
github.com/labstack/armor/vendor/github.com/valyala/bytebufferpool
github.com/labstack/armor/vendor/github.com/mattn/go-colorable
github.com/labstack/armor/vendor/github.com/valyala/fasttemplate
github.com/labstack/armor/vendor/github.com/labstack/gommon/color
github.com/labstack/armor/vendor/github.com/Knetic/govaluate
github.com/labstack/armor/vendor/github.com/labstack/gommon/log
github.com/labstack/armor/vendor/github.com/golang/protobuf/proto
github.com/labstack/armor/vendor/github.com/labstack/gommon/bytes
github.com/labstack/armor/vendor/github.com/labstack/gommon/random
github.com/labstack/armor/vendor/github.com/mitchellh/mapstructure
github.com/labstack/armor/util
github.com/labstack/armor/vendor/github.com/mitchellh/go-homedir
crypto/x509
vendor/golang_org/x/net/lex/httplex
vendor/golang_org/x/net/proxy
net/textproto
mime/multipart
crypto/tls
github.com/labstack/armor/vendor/github.com/dgrijalva/jwt-go
net/http/httptrace
net/http
github.com/labstack/armor/vendor/golang.org/x/crypto/acme
net/http/httputil
github.com/labstack/armor/vendor/golang.org/x/crypto/acme/autocert
github.com/labstack/armor/vendor/github.com/labstack/echo
github.com/labstack/armor
github.com/labstack/armor/vendor/github.com/labstack/echo/middleware
github.com/labstack/armor/plugin
github.com/labstack/armor/http
command-line-arguments
Removing intermediate container 0e3c19dc9727
 ---> 7657ebc0d185
Step 6/10 : FROM scratch
 ---> 
Step 7/10 : WORKDIR /
Removing intermediate container 98dacab1196c
 ---> 11059256dd36
Step 8/10 : COPY --from=builder /go/bin/armor /armor
 ---> c5314aff8317
Step 9/10 : COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
 ---> c5490e105640
Step 10/10 : ENTRYPOINT ["/armor"]
 ---> Running in 7f4a0c18fa52
Removing intermediate container 7f4a0c18fa52
 ---> 43c99df892a0
Successfully built 43c99df892a0
Successfully tagged armor-pull:latest
```

File: Dockerfile (hunk 1 above)
Comment (alexandreStein) [reply Dockerfile]: I find a good workaround by using an intermediate container.

This looks good to me.

Comment (im-kulikov) [reply Dockerfile]: it's because you copy `package.tar` to your container.. not extract, only copied

Comment (im-kulikov) [reply Dockerfile]: You must do something like this:

```
# Build image
FROM golang:alpine AS builder
MAINTAINER Vishal Rana <vr@labstack.com>

COPY . /go/src/github.com/labstack/armor

WORKDIR /go/src/github.com/labstack/armor

RUN set -x \
    && export CGO_ENABLED=0 \
    && go build -v -o /go/bin/armor cmd/armor/main.go

RUN mkdir -p /etc/armor/certs

COPY ./config.yaml /etc/armor/config.yaml
COPY ./_fixture /var/www/test/

# Executable image
FROM scratch

WORKDIR /

COPY --from=builder /go/bin/armor /armor
COPY --from=builder /etc/armor /etc/armor
COPY --from=builder /var/www /var/www
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/

ENTRYPOINT ["/armor", "-c", "/etc/armor/config.yaml"]
```

Comment (alexandreStein) [reply Dockerfile]: Sorry, I'm just missing up...
This should work.

Comment (im-kulikov) [reply Dockerfile]: nice.. waiting for review @vishr @alexandreStein @SkyrisBactera 🙂

**Comment by vishr** [comment]
@im-kulikov Thanks for your effort in minimizing the container size. The new size I am getting is 9.87 MB in comparison to 14.3 MB. As @alexandreStein mentioned 14 MB is a decent size specially if you are getting shell, tools and a package manager to install anything on the container for debugging. Please let me know your thoughts.

**Comment by im-kulikov** [comment]
@vishr `scratch` is empty container without dependencies
we don't need golang or other libs after building container
more dependencies - more potential vulnerabilities

**Comment by im-kulikov** [comment]
on the other hand, what are you going debajit?
alpine:edge unstable version, that is, there is no guarantee that today's packages will be available tomorrow

**Comment by vishr** [comment]
@im-kulikov We are compiling locally and only putting the binary in an alpine image, there is no golang deps or other libraries in the container. Ideally we should build as you are doing in the container and I am okay with that.

What is debajit?

Here is what I would suggest:

- Use golang as builder (with version) - OK
- Replace alpine (with version) for executable image

Let me know your thoughts.

**Comment by im-kulikov** [comment]
@vishr oh, sorry.. mistake “debajit” == “debug”

**Comment by im-kulikov** [comment]
what is the meaning of debug in the container for production.

**Comment by vishr** [comment]
For instance, I just want to see my config file or run some command, without a shell I can't move a bit. Alpine is just adding 4 MB and I think it is OK.

**Comment by im-kulikov** [comment]
maybe it makes sense to make two Dockerfile? one for production, without any external dependencies.. what do you think?

**Comment by vishr** [comment]
It will be complicated. What external dependencies are you talking about?

**Comment by im-kulikov** [comment]
GNU/Linux tools..

**Comment by vishr** [comment]
I doubt as most of the people will definitely need shell and basic tools even in production.

**Comment by im-kulikov** [comment]
these binaries are not needed in daily work..
in my work and deploy I use mount inside the container. what should change, in theory should be outside the container (IMHO)

**Comment by vishr** [comment]
If you think from a common user's perspective - they will need them. For use cases like yours I think it is better to grab a binary from the releases and build your own image.

**Comment by im-kulikov** [comment]
hm..

**Comment by im-kulikov** [comment]
okay, what I can do for merge?

**Comment by vishr** [comment]
- Use a version in golang builder
- Replace scratch with alpine:3.7?
- If you look at Makefile, we do cross compile - it will be good to include those steps in the builder image by mounting build directory

**Comment by vishr** [comment]
@im-kulikov I think you got it wrong. You previous PR was okay, I just wanted to replace scratch with alpine:3.7 and use a golang builder with some tag.

**Comment by vishr** [comment]
For now you can leave it as is. I will pull it and do the fix.

**Comment by im-kulikov** [comment]
@vishr ok

**Comment by im-kulikov** [comment]
@vishr mb rename PR to `Cleanup and stable alpine`?

**Comment by vishr** [comment]
Yes that is fine.

**Comment by im-kulikov** [comment]
fixed

**Pull request merged**